File: zadanie3-2.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int done;
    struct node *left;
    struct node *right;
};

struct node* createNode(int done) {
    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    newNode->done = done;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

struct node* root = NULL;
int najlepsie = 10000;

void rekurzia_na_strome(char slovo[], int i, struct node* actual_node, int pocet_iter){
    if (slovo[0] == "0") {
        return;
    }
    if (i == strlen(slovo)-1 )
    {
        if (pocet_iter < najlepsie)
        {
            najlepsie = pocet_iter +1;  
        }
        
            
    }

    int j = i;

    while (j < strlen(slovo) - 1)
    {
        if (slovo[j+1] == '0')
        {
            if (actual_node->left)
            {
                actual_node = actual_node->left;
            }
            else
            {
                break;
            }
        }
        else
        {
           if (actual_node->right)
            {
                actual_node = actual_node->right;
                if (actual_node->done)
                {                    
                    if (j+2 == strlen(slovo))
                    {
                        if (pocet_iter < najlepsie)
                        {
                            najlepsie = pocet_iter+1;  
                            break;
                        }
                    }
                    
                    rekurzia_na_strome(slovo, j+2, root,pocet_iter + 1);
                }
                
            }
            else
            {
                if (actual_node->done)
                {
                    rekurzia_na_strome(slovo, j+1, root,pocet_iter + 1);
                }
                
                break;
            }
        }
        
        j++;
    }
}

void reverseString(char str[]) {
    int length = strlen(str);
    int i, j;
    char temp;
    for (i = 0, j = length - 1; i < j; i++, j--) {
        temp = str[i];
        str[i] = str[j];
        str[j] = temp;
    }
}

int main() {
    root = createNode(1);
    for (int i = 1; i < 22; i++)
    {
        struct node* actual_node = root;
        long long new_num = (long long) pow(5,i)/2;
        int j = 0;
        while (new_num > 0) {
            int next_num = new_num % 2;


            if (next_num == 0){
                if (!actual_node->left){
                    actual_node->left = createNode(0);
                }
                actual_node = actual_node->left;
            }
            else{
                if (!actual_node->right){
                    if (j == (2*i)-1)
                    {
                        actual_node->right = createNode(1);
                    }
                    else
                    {
                        actual_node->right = createNode(0);
                    }                    
                }
                actual_node = actual_node->right;
            }
            new_num /= 2;
            j++;
            
        }
    }
    char vstup[50];
    char ch;
    while(2 == scanf("%s%c", &vstup, &ch)){ 
        if (ch == '\n')
        {
            reverseString(vstup);
            rekurzia_na_strome(vstup,0,root,0);

            if (najlepsie == 10000)
            {
                printf("-1\n");
            }
            else
            {
                printf("%d",najlepsie);
            }
            najlepsie = 10000;
        }
    }
}

File: zadanie3-2.py
class TreeNode:
    def __init__(self, done):
        self.done = done
        self.one = None
        self.zer = None

    def add_one(self, done):
        self.one = TreeNode(done)

    def add_zer(self, done):
        self.zer = TreeNode(done)

root = TreeNode(True)


mocniny_5 = []
cislo1 = "1"
cislo2 = 1


while True:
    actual_node = root
    i = 0
    mocniny_5.append(cislo1)
    cislo1 = bin(pow(5,cislo2))[2:]
    if len(cislo1) > 50:
        break
    print((cislo1[::-1])[1:])
    while i < len(cislo1) - 1:
        next_num = cislo1[len(cislo1)-i-2]


        if next_num == "0":
            if not actual_node.zer: 
                if len(cislo1)-2-i == 0:
                    actual_node.add_zer(True)
                else:
                    actual_node.add_zer(False)
            actual_node = actual_node.zer



        else:
            if not actual_node.one: 
                if len(cislo1)-2-i == 0:
                    print(i)
                    actual_node.add_one(True)
                else:
                    actual_node.add_one(False)
            actual_node = actual_node.one
            
        i+=1
            
    cislo2 += 1



najlepsie = 10000

def rekurzia_na_strome(slovo,i,actual_node,pocet_iter,nas_string):
    global najlepsie
    if slovo[i] == "0":
        return 
    
    if i == len(slovo)-1:
        if pocet_iter < najlepsie:
            najlepsie = pocet_iter+1
    
    j = i
    while j < len(slovo)-1:
        if slovo[j+1] == '0':
            if actual_node.zer:
                actual_node = actual_node.zer
                nas_string += "0"
            else:
                break 
        else:
            if actual_node.one:
                actual_node = actual_node.one
                nas_string += "1"
                if actual_node.done:
                    if j+2 == len(slovo):
                        if pocet_iter < najlepsie:
                            najlepsie = pocet_iter+1
                            break
                    rekurzia_na_strome(slovo,j+2,root,pocet_iter+1,nas_string)
            else:
                if actual_node.done:
                    rekurzia_na_strome(slovo,j+1,root,pocet_iter+1,nas_string)
                break

        j+=1
    
        
rekurzia_na_strome(str(1010101)[::-1],0,root,0,"1")


if najlepsie == 10000:
    print(-1)
else:
    print(najlepsie)


